src/utils/authMiddlewares.js

```diff
@@ -0,0 +1,156 @@
+const jwtExpress = require('express-jwt');
+const jwkToPem = require('jwk-to-pem');
+const util = require('util');
+const config = require('../config');
+const CacheSingleton = require('../cache');
+const { CacheMissError } = require('../cache/cache-utils');
+const { UnauthorizedError, UnauthenticedError } = require('./errors');
+const ExperimentService = require('../api/route-services/experiment');
+
+const experimentService = new ExperimentService();
+
+/**
+ * Authentication middleware for Express. Returns a middleware that
+ * can be used in the API to authenticate Cognito-issued JWTs.
+ */
+const authenticationMiddlewareExpress = async (app) => {
+  app.set('keys', {});
+  const poolId = await config.awsUserPoolIdPromise;
+
+  return jwtExpress({
+    algorithms: ['RS256'],
+    credentialsRequired: false,
+    issuer: `https://cognito-idp.${config.awsRegion}.amazonaws.com/${poolId}`,
+    secret: (req, payload, done) => {
+      const token = req.headers.authorization.split(' ')[1];
+      const [jwtHeaderRaw] = token.split('.');
+      const { kid } = JSON.parse(Buffer.from(jwtHeaderRaw, 'base64').toString('ascii'));
+
+      // Get the issuer from the JWT claim.
+      const { iss } = payload;
+
+      if (!iss.endsWith(poolId)) {
+        done('Issuer does not correspond to the correct environment.');
+      }
+
+      // If we have already seen the key this message was signed by,
+      // return immediately.
+      const existingPem = app.get('keys').kid;
+
+      if (existingPem) {
+        done(null, existingPem);
+        return;
+      }
+
+      // Otherwise, find the appropriate key for the issuer.
+      fetch(`${iss}/.well-known/jwks.json`).then((res) => res.json()).then(({ keys }) => {
+        const secret = keys.find((key) => key.kid === kid);
+        const pem = jwkToPem(secret);
+
+        app.set('keys', { ...app.get('keys'), kid: pem });
+        done(null, pem);
+      });
+    },
+  });
+};
+
+/**
+ * Authentication middleware for Socket.IO requests. Resolves with
+ * the JWT claim if the authentication was successful, or rejects with
+ * the error otherwise.
+ *
+ * @param {*} authHeader The bearer-encoded JWT token.
+ * @returns Promise that resolves or rejects based on authentication status.
+ */
+const authenticationMiddlewareSocketIO = async (authHeader) => {
+  const poolId = await config.awsUserPoolIdPromise;
+  const cache = CacheSingleton.get();
+
+  const issuer = `https://cognito-idp.${config.awsRegion}.amazonaws.com/${poolId}`;
+  const token = authHeader.split(' ')[1];
+  const jwtVerify = util.promisify(jwt.verify);
+
+  const result = await jwtVerify(
+    token,
+    (header, callback) => {
+      const { kid } = header;
+
+      cache.get(kid)
+        .then((pem) => callback(null, pem))
+        .catch((e) => {
+          if (!(e instanceof CacheMissError)) {
+            throw e;
+          }
+
+          fetch(`${issuer}/.well-known/jwks.json`).then((res) => res.json()).then(({ keys }) => {
```

Comment:
Does it make sense to perhaps factor out into a different function the fetching of the key? Since it is being used both here and in the other middleware.

Reply:
It would be good however the functions themselves are quite different. I could factor out the actual fetch call, but even then because we have two different types of caching being used I’ve found it hard to abstract it away properly

Reply:
I was thinking mainly handling the lines 90, 91 and 92 in the new function and returning a promise to separate caching from fetching, but if it has the potential to cause issues I think its ok like this.